src/community/features-templating/features-templating-rest/src/main/java/org/geoserver/featurestemplating/rest/TemplateRestController.java

```diff
@@ -0,0 +1,613 @@
+        try {
+            byte[] rawData = IOUtils.toByteArray(inputStream);
+            String content = new String(rawData, Charset.defaultCharset());
+            TemplateFileManager.get().saveTemplateFile(info, content);
+            TemplateInfoDAO.get().saveOrUpdate(info);
+        } catch (IOException e) {
+            throw new RestException(
+                    "Error while writing the template", HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    @DeleteMapping(
+        value = {
+            "/featurestemplates/{templateName}",
+            "/featuretypes/{featureType}/featurestemplates/{templateName}",
+            "/workspaces/{ws}/featurestemplates/{templateName}"
+        },
+        produces = MediaType.TEXT_PLAIN_VALUE
+    )
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void deleteTemplate(
+            @PathVariable(required = false) String featureType,
+            @PathVariable(required = false) String ws,
+            @PathVariable String templateName) {
+
+        if (ws != null) {
+            checkWorkspaceName(ws);
+            checkFullAdminRequired(ws);
+        }
+        FeatureTypeInfo ft;
+        if (featureType != null) {
+            ft = checkFeatureType(featureType);
+            ws = ft.getStore().getWorkspace().getName();
+            featureType = ft.getName();
+        }
+        String fullName = buildFullName(ws, featureType, templateName);
+        TemplateInfoDAO dao = TemplateInfoDAO.get();
+        TemplateInfo info = dao.findByFullName(fullName);
+        TemplateFileManager.get().delete(info);
+        dao.delete(info);
+        LOGGER.info("Deleted template with name " + info.getFullName());
+    }
+
+    @GetMapping(
+        value = {
+            "/featurestemplates/{templateName}",
+            "/featuretypes/{featureType}/featurestemplates/{templateName}",
+            "/workspaces/{ws}/featurestemplates/{templateName}"
+        },
+        produces = {
+            MediaType.APPLICATION_XML_VALUE,
+            MediaType.APPLICATION_JSON_VALUE,
+            MediaType.APPLICATION_XHTML_XML_VALUE,
+        }
+    )
+    public void templateGet(
+            @PathVariable(required = false) String ws,
+            @PathVariable(required = false) String featureType,
+            @PathVariable String templateName,
+            HttpServletResponse response) {
+
+        if (ws != null) {
+            checkWorkspaceName(ws);
+            checkFullAdminRequired(ws);
+        }
+        FeatureTypeInfo ft;
+        if (featureType != null) {
+            ft = checkFeatureType(featureType);
+            ws = ft.getStore().getWorkspace().getName();
+            featureType = ft.getName();
+        }
+        String fullName = buildFullName(ws, featureType, templateName);
+        TemplateInfo info = TemplateInfoDAO.get().findByFullName(fullName);
+        Resource resource = TemplateFileManager.get().getTemplateResource(info);
```

Comment:
Yes this could have been made more generic. Currently the template content is handle as a FileSystemResource also at the level of TemplateResourceLoader since the plug-in born. That's why the TemplateFileManager plays directly with File objects.
Anyway this will likely be modified in the future to support other storage types.